src/contracts/AbstractARM.sol

```diff
@@ -726,4 +771,132 @@ abstract contract AbstractARM is OwnableOperable, ERC20Upgradeable {
+            require(market != address(0), "ARM: invalid market");
+            require(!supportedMarkets[market], "ARM: market already supported");
+            require(IERC4626(market).asset() == liquidityAsset, "ARM: invalid market asset");
+
+            supportedMarkets[market] = true;
+
+            emit MarketAdded(market);
+        }
+    }
+
+    /// @notice Owner removes a supported lending market from the ARM.
+    /// This can not be the active market.
+    /// @param _market The address of the lending market to remove
+    function removeMarket(address _market) external onlyOwner {
+        require(_market != address(0), "ARM: invalid market");
+        require(supportedMarkets[_market], "ARM: market not supported");
+        require(_market != activeMarket, "ARM: market in active");
+
+        supportedMarkets[_market] = false;
+
+        emit MarketRemoved(_market);
+    }
+
+    /// @notice set a new active lending market for the ARM.
+    /// This can be set to address(0) to disable the use of a lending market.
+    function setActiveMarket(address _market) external onlyOperatorOrOwner {
+        require(_market == address(0) || supportedMarkets[_market], "ARM: market not supported");
+        // Read once from storage to save gas and make it clear this is the previous active market
+        address previousActiveMarket = activeMarket;
+        require(previousActiveMarket != _market, "ARM: already active market");
+
+        if (previousActiveMarket != address(0)) {
+            // Redeem all shares from the previous active lending market.
+            // balanceOf is used instead of maxRedeem to ensure all shares are redeemed.
+            // maxRedeem can return a smaller amount of shares than balanceOf if the market is highly utilized.
+            uint256 shares = IERC4626(previousActiveMarket).balanceOf(address(this));
+            // This could fail if the market has high utilization
+            if (shares > minSharesToRedeem) {
+                IERC4626(previousActiveMarket).redeem(shares, address(this), address(this));
+            }
+        }
+
+        activeMarket = _market;
+
+        emit ActiveMarketUpdated(_market);
+
+        // Exit if no new active market
+        if (_market == address(0)) return;
+
+        _allocate();
+    }
+
+    function allocate() external {
+        require(activeMarket != address(0), "ARM: no active market");
+
+        _allocate();
+    }
+
+    function _allocate() internal {
+        (uint256 availableAssets, uint256 outstandingWithdrawals) = _availableAssets();
+        if (availableAssets == 0) return;
+
+        int256 armLiquidity = SafeCast.toInt256(IERC20(liquidityAsset).balanceOf(address(this)))
+            - SafeCast.toInt256(outstandingWithdrawals);
+        uint256 targetArmLiquidity = availableAssets * armBuffer / 1e18;
+
+        int256 liquidityDelta = armLiquidity - SafeCast.toInt256(targetArmLiquidity);
+
+        if (liquidityDelta > 0) {
```

Comment:
Perhaps we should have some kind of hysteresis on `liquidityDelta`, rather just allocating when it's not perfect.

Well behaved 4626's round against the protocol, so it feels like it would be be possible to get into a situation where someone could call allocate() over and over again, each time removing a rounding error of value.

Just feels a little safer and more efficient in general to have something like a 1% threshold required on either side.

Reply:
This is a good point. We can have something like

```
if (liquidityDelta > minLiquidityThreshold) {
```

I was thinking we could reuse `minSharesToRedeem` or `MIN_TOTAL_SUPPLY` but those are of type `uint256`. And its better to not mix concerns.

Reply:
I've added `allocateThreshold` immutable in PR https://github.com/OriginProtocol/arm-oeth/pull/84

```Solidity
// The allocateThreshold prevents the ARM from constantly depositing and withdrawing if there are rounding issues
if (liquidityDelta > allocateThreshold) {
    // We have too much liquidity in the ARM, we need to deposit some to the active lending market

    uint256 depositAmount = SafeCast.toUint256(liquidityDelta);

    IERC20(liquidityAsset).approve(activeMarket, depositAmount);
    IERC4626(activeMarket).deposit(depositAmount, address(this));
} else if (liquidityDelta < 0) {
```